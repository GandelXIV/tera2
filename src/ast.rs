use std::collections::HashMap;
use std::fmt;

use crate::lexer::Operator;

/// An expression is the node found in variable block, kwargs and conditions.
#[derive(Clone, Debug, PartialEq)]
#[allow(missing_docs)]
pub enum Expression {
    String(String),
    Int(i64),
    Float(f64),
    Bool(bool),
    Ident(String),
    Array(Vec<Expression>),
    // name, args
    Test(String, Vec<Expression>),
    // namespace, name, kwargs
    MacroCall(String, String, HashMap<String, Expression>),
    // name, kwargs
    Function(String, HashMap<String, Expression>),
    Expr(Operator, Vec<Expression>),
}

impl Expression {
    pub(crate) fn as_array(self) -> Vec<Expression> {
        match self {
            Expression::Array(vals) => vals,
            _ => panic!("Called as_array on a non array value"),
        }
    }
}

impl fmt::Display for Expression {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Expression::*;

        match self {
            String(i) => write!(f, "'{}'", i),
            Int(i) => write!(f, "{}", i),
            Float(i) => write!(f, "{}", i),
            Ident(i) => write!(f, "{}", i),
            Bool(i) => write!(f, "{}", i),
            Array(vals) => {
                write!(f, "[")?;
                for (i, s) in vals.iter().enumerate() {
                    if i == vals.len() - 1 {
                        write!(f, "{}", s)?
                    } else {
                        write!(f, "{}, ", s)?
                    }
                }
                write!(f, "]")
            }
            Test(name, rest) => {
                write!(f, "{}", name)?;

                if !rest.is_empty() {
                    write!(f, "{{",)?;
                    for (i, s) in rest.iter().enumerate() {
                        if i == rest.len() - 1 {
                            write!(f, "{}", s)?
                        } else {
                            write!(f, "{}, ", s)?
                        }
                    }
                    write!(f, "}}",)?;
                }
                Ok(())
            }
            MacroCall(namespace, name, kwargs) => {
                write!(f, "{}::{}", namespace, name)?;
                write!(f, "{{",)?;
                let mut keys = kwargs.keys().collect::<Vec<_>>();
                keys.sort();
                for (i, k) in keys.iter().enumerate() {
                    if i == kwargs.len() - 1 {
                        write!(f, "{}={}", k, kwargs[*k])?
                    } else {
                        write!(f, "{}={}, ", k, kwargs[*k])?
                    }
                }
                write!(f, "}}",)?;
                Ok(())
            }
            Function(name, kwargs) => {
                write!(f, "{}", name)?;
                write!(f, "{{",)?;
                let mut keys = kwargs.keys().collect::<Vec<_>>();
                keys.sort();
                for (i, k) in keys.iter().enumerate() {
                    if i == kwargs.len() - 1 {
                        write!(f, "{}={}", k, kwargs[*k])?
                    } else {
                        write!(f, "{}={}, ", k, kwargs[*k])?
                    }
                }
                write!(f, "}}",)?;
                Ok(())
            }
            Expr(op, rest) => {
                write!(f, "({}", op)?;
                for s in rest {
                    write!(f, " {}", s)?
                }
                write!(f, ")")
            }
        }
    }
}

/// Set a variable in the context `{% set val = "hey" %}`
#[derive(Clone, Debug, PartialEq)]
pub struct Set {
    /// The name for that value in the context
    pub key: String,
    /// The value to assign
    pub value: Expression,
    /// Whether we want to set the variable globally or locally
    /// global_set is only useful in loops
    pub global: bool,
}

/// All Tera nodes that can be encountered
#[derive(Clone, Debug, PartialEq)]
pub enum Node {
    Text(String),
    Expression(Expression),
    Set(Set),
    Include {
        files: Vec<String>,
        ignore_missing: bool,
    },
}
