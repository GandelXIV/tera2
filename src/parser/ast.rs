use std::collections::HashMap;
use std::fmt;

use crate::parser::lexer::Operator;
use crate::utils::Spanned;

/// The various possible literals in variable blocks/squared brackets
#[derive(Clone, Debug, PartialEq)]
#[allow(missing_docs)]
pub enum Literal {
    Str(String),
    Integer(i64),
    Float(f64),
    Bool(bool),
}

/// An expression is the node found in variable block, kwargs and conditions.
#[derive(Clone, Debug, PartialEq)]
#[allow(missing_docs)]
pub enum Expression {
    Str(String),
    Integer(i64),
    Float(f64),
    Bool(bool),
    Ident(String),
    Array(Vec<SpannedExpression>),
    Test(Test),
    MacroCall(MacroCall),
    FunctionCall(FunctionCall),
    Expr(Operator, Vec<SpannedExpression>),
}

pub type SpannedExpression = Spanned<Expression>;

#[derive(Clone, Debug, PartialEq)]
pub struct Test {
    pub name: String,
    pub args: Vec<SpannedExpression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct MacroCall {
    pub namespace: String,
    pub name: String,
    pub kwargs: HashMap<String, SpannedExpression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct FunctionCall {
    pub name: String,
    pub kwargs: HashMap<String, SpannedExpression>,
}

impl SpannedExpression {
    pub(crate) fn can_be_iterated_on(&self) -> bool {
        use Expression::*;
        matches!(
            self.node,
            Str(_) | Ident(_) | Array(_) | FunctionCall(_) | Expr(_, _)
        )
    }
}

impl fmt::Display for SpannedExpression {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Expression::*;

        match &self.node {
            Str(i) => write!(f, "'{}'", i),
            Integer(i) => write!(f, "{}", i),
            Float(i) => write!(f, "{}", i),
            Ident(i) => write!(f, "{}", i),
            Bool(i) => write!(f, "{}", i),
            Array(vals) => {
                write!(f, "[")?;
                for (i, s) in vals.iter().enumerate() {
                    if i == vals.len() - 1 {
                        write!(f, "{}", s)?
                    } else {
                        write!(f, "{}, ", s)?
                    }
                }
                write!(f, "]")
            }
            Test(t) => {
                write!(f, "{}", t.name)?;

                if !t.args.is_empty() {
                    write!(f, "{{",)?;
                    for (i, s) in t.args.iter().enumerate() {
                        if i == t.args.len() - 1 {
                            write!(f, "{}", s)?
                        } else {
                            write!(f, "{}, ", s)?
                        }
                    }
                    write!(f, "}}",)?;
                }
                Ok(())
            }
            MacroCall(mc) => {
                write!(f, "{}::{}", mc.namespace, mc.name)?;
                write!(f, "{{",)?;
                let mut keys = mc.kwargs.keys().collect::<Vec<_>>();
                keys.sort();
                for (i, k) in keys.iter().enumerate() {
                    if i == mc.kwargs.len() - 1 {
                        write!(f, "{}={}", k, mc.kwargs[*k])?
                    } else {
                        write!(f, "{}={}, ", k, mc.kwargs[*k])?
                    }
                }
                write!(f, "}}",)?;
                Ok(())
            }
            FunctionCall(fc) => {
                write!(f, "{}", fc.name)?;
                write!(f, "{{",)?;
                let mut keys = fc.kwargs.keys().collect::<Vec<_>>();
                keys.sort();
                for (i, k) in keys.iter().enumerate() {
                    if i == fc.kwargs.len() - 1 {
                        write!(f, "{}={}", k, fc.kwargs[*k])?
                    } else {
                        write!(f, "{}={}, ", k, fc.kwargs[*k])?
                    }
                }
                write!(f, "}}",)?;
                Ok(())
            }
            Expr(op, rest) => {
                write!(f, "({}", op)?;
                for s in rest {
                    write!(f, " {}", s)?
                }
                write!(f, ")")
            }
        }
    }
}

/// Set a variable in the context `{% set val = "hey" %}`
#[derive(Clone, Debug, PartialEq)]
pub struct Set {
    /// The name for that value in the context
    pub key: String,
    /// The value to assign
    pub value: SpannedExpression,
    /// Whether we want to set the variable globally or locally
    /// global_set is only useful in loops
    pub global: bool,
}

/// A block definition
#[derive(Clone, Debug, PartialEq)]
pub struct Block {
    /// The block name
    pub name: String,
    /// The block content
    pub body: Vec<Node>,
}

/// An if/elif/else condition with their respective body
#[derive(Clone, Debug, PartialEq)]
pub struct If {
    /// First item if the if, all the ones after are elif
    pub conditions: Vec<(SpannedExpression, Vec<Node>)>,
    /// The optional `else` block
    pub otherwise: Vec<Node>,
}

/// A filter section node `{{ filter name(param="value") }} content {{ endfilter }}`
#[derive(Clone, Debug, PartialEq)]
pub struct FilterSection {
    pub name: String,
    pub kwargs: HashMap<String, SpannedExpression>,
    /// The filter body
    pub body: Vec<Node>,
}

/// A Macro definition `{% macro hello() %}...{% endmacro %}`
#[derive(Clone, Debug, PartialEq)]
pub struct MacroDefinition {
    pub name: String,
    /// The args for that macro: name -> optional default value
    pub kwargs: HashMap<String, Option<SpannedExpression>>,
    pub body: Vec<Node>,
}

/// A forloop: can be over values or key/values
#[derive(Clone, Debug, PartialEq)]
pub struct ForLoop {
    /// Name of the key in the loop (only when iterating on map-like objects)
    pub key: Option<String>,
    /// Name of the local variable for the value in the loop
    pub value: String,
    /// Expression being iterated on
    pub container: SpannedExpression,
    /// What's in the forloop itself
    pub body: Vec<Node>,
    /// The body to execute in case of an empty object in the `{% for .. %}{% else %}{% endfor %}` construct
    pub otherwise: Vec<Node>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Include {
    pub files: Vec<String>,
    pub ignore_missing: bool,
}

/// All Tera nodes that can be encountered
#[derive(Clone, Debug, PartialEq)]
pub enum Node {
    Text(String),
    VariableBlock(SpannedExpression),
    Set(Set),
    Raw(String),
    Include(Include),
    Block(Block),
    Super,
    If(If),
    ForLoop(ForLoop),
    Continue,
    Break,
    FilterSection(FilterSection),
}
