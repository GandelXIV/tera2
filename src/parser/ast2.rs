use std::collections::HashMap;
use std::fmt;

use crate::utils::Spanned;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum UnaryOperator {
    Not,
    Minus,
}

impl fmt::Display for UnaryOperator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use UnaryOperator::*;

        let val = match self {
            Minus => "-",
            Not => "not",
        };
        write!(f, "{}", val)
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum BinaryOperator {
    // math
    Mul,
    Div,
    Mod,
    Plus,
    Minus,
    FloorDiv,
    Power,

    // comparison
    LessThan,
    GreaterThan,
    LessThanOrEqual,
    GreaterThanOrEqual,
    Equal,
    NotEqual,

    // rest
    And,
    Or,
    StrConcat,
    In,
    Is,
    Pipe,
}

impl fmt::Display for BinaryOperator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use BinaryOperator::*;

        let val = match self {
            Mul => "*",
            Power => "**",
            Div => "/",
            FloorDiv => "//",
            Mod => "%",
            Plus => "+",
            Minus => "-",
            LessThan => "<",
            GreaterThan => ">",
            LessThanOrEqual => "<=",
            GreaterThanOrEqual => ">=",
            Equal => "==",
            NotEqual => "!=",
            And => "and",
            Or => "or",
            In => "in",
            Is => "is",
            StrConcat => "~",
            Pipe => "|",
        };
        write!(f, "{}", val)
    }
}

// TODO: handle my_data.blob[val].hey at compile time -> Eg no ident

/// An expression is the node found in variable block, kwargs and conditions.
#[derive(Clone, PartialEq)]
#[allow(missing_docs)]
pub enum Expression {
    Str(Spanned<String>),
    Integer(Spanned<i64>),
    Float(Spanned<f64>),
    Bool(Spanned<bool>),
    Ident(Spanned<String>),
    Array(Spanned<Array>),
    Test(Spanned<Test>),
    MacroCall(Spanned<MacroCall>),
    FunctionCall(Spanned<FunctionCall>),
    UnaryOperation(Spanned<UnaryOperation>),
    BinaryOperation(Spanned<BinaryOperation>),
}

#[derive(Clone, PartialEq)]
pub struct UnaryOperation {
    pub op: UnaryOperator,
    pub expr: Expression,
}

impl fmt::Debug for UnaryOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({} {:?})", self.op, self.expr)
    }
}

#[derive(Clone, PartialEq)]
pub struct BinaryOperation {
    pub op: BinaryOperator,
    pub left: Expression,
    pub right: Expression,
}

impl fmt::Debug for BinaryOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({} {:?} {:?})", self.op, self.left, self.right)
    }
}

#[derive(Clone, PartialEq)]
pub struct Array {
    pub items: Vec<Expression>,
}

impl fmt::Debug for Array {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[")?;
        for (i, s) in self.items.iter().enumerate() {
            if i == self.items.len() - 1 {
                write!(f, "{:?}", s)?
            } else {
                write!(f, "{:?}, ", s)?
            }
        }
        write!(f, "]")
    }
}

#[derive(Clone, PartialEq)]
pub struct Test {
    pub name: String,
    pub args: Vec<Expression>,
}

impl fmt::Debug for Test {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)?;

        if !self.args.is_empty() {
            write!(f, "{{",)?;
            for (i, s) in self.args.iter().enumerate() {
                if i == self.args.len() - 1 {
                    write!(f, "{:?}", s)?
                } else {
                    write!(f, "{:?}, ", s)?
                }
            }
            write!(f, "}}",)?;
        }
        Ok(())
    }
}

#[derive(Clone, PartialEq)]
pub struct MacroCall {
    pub namespace: String,
    pub name: String,
    pub kwargs: HashMap<String, Expression>,
}

impl fmt::Debug for MacroCall {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}::{}", self.namespace, self.name)?;
        write!(f, "{{",)?;
        let mut keys = self.kwargs.keys().collect::<Vec<_>>();
        keys.sort();
        for (i, k) in keys.iter().enumerate() {
            if i == self.kwargs.len() - 1 {
                write!(f, "{}={:?}", k, self.kwargs[*k])?
            } else {
                write!(f, "{}={:?}, ", k, self.kwargs[*k])?
            }
        }
        write!(f, "}}",)
    }
}

#[derive(Clone, PartialEq)]
pub struct FunctionCall {
    pub name: String,
    pub kwargs: HashMap<String, Expression>,
}

impl fmt::Debug for FunctionCall {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)?;
        write!(f, "{{",)?;
        let mut keys = self.kwargs.keys().collect::<Vec<_>>();
        keys.sort();
        for (i, k) in keys.iter().enumerate() {
            if i == self.kwargs.len() - 1 {
                write!(f, "{}={:?}", k, self.kwargs[*k])?
            } else {
                write!(f, "{}={:?}, ", k, self.kwargs[*k])?
            }
        }
        write!(f, "}}",)
    }
}

impl fmt::Debug for Expression {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Expression::*;

        match self {
            Str(i) => write!(f, "'{}'", **i),
            Integer(i) => write!(f, "{}", **i),
            Float(i) => write!(f, "{}", **i),
            Ident(i) => write!(f, "{}", **i),
            Bool(i) => write!(f, "{}", **i),
            Array(i) => write!(f, "{:?}", **i),
            Test(i) => write!(f, "{:?}", **i),
            MacroCall(i) => write!(f, "{:?}", **i),
            FunctionCall(i) => write!(f, "{:?}", **i),
            UnaryOperation(i) => write!(f, "{:?}", **i),
            BinaryOperation(i) => write!(f, "{:?}", **i),
        }
    }
}

// /// Set a variable in the context `{% set val = "hey" %}`
// #[derive(Clone, Debug, PartialEq)]
// pub struct Set {
//     /// The name for that value in the context
//     pub key: String,
//     /// The value to assign
//     pub value: SpannedExpression,
//     /// Whether we want to set the variable globally or locally
//     /// global_set is only useful in loops
//     pub global: bool,
// }
//
// /// A block definition
// #[derive(Clone, Debug, PartialEq)]
// pub struct Block {
//     /// The block name
//     pub name: String,
//     /// The block content
//     pub body: Vec<Node>,
// }
//
// /// An if/elif/else condition with their respective body
// #[derive(Clone, Debug, PartialEq)]
// pub struct If {
//     /// First item if the if, all the ones after are elif
//     pub conditions: Vec<(SpannedExpression, Vec<Node>)>,
//     /// The optional `else` block
//     pub otherwise: Vec<Node>,
// }
//
// /// A filter section node `{{ filter name(param="value") }} content {{ endfilter }}`
// #[derive(Clone, Debug, PartialEq)]
// pub struct FilterSection {
//     pub name: String,
//     pub kwargs: HashMap<String, SpannedExpression>,
//     /// The filter body
//     pub body: Vec<Node>,
// }
//
// /// A Macro definition `{% macro hello() %}...{% endmacro %}`
// #[derive(Clone, Debug, PartialEq)]
// pub struct MacroDefinition {
//     pub name: String,
//     /// The args for that macro: name -> optional default value
//     pub kwargs: HashMap<String, Option<SpannedExpression>>,
//     pub body: Vec<Node>,
// }
//
// /// A forloop: can be over values or key/values
// #[derive(Clone, Debug, PartialEq)]
// pub struct ForLoop {
//     /// Name of the key in the loop (only when iterating on map-like objects)
//     pub key: Option<String>,
//     /// Name of the local variable for the value in the loop
//     pub value: String,
//     /// Expression being iterated on
//     pub container: SpannedExpression,
//     /// What's in the forloop itself
//     pub body: Vec<Node>,
//     /// The body to execute in case of an empty object in the `{% for .. %}{% else %}{% endfor %}` construct
//     pub otherwise: Vec<Node>,
// }
//
// #[derive(Clone, Debug, PartialEq)]
// pub struct Include {
//     pub files: Vec<String>,
//     pub ignore_missing: bool,
// }
//
// /// All Tera nodes that can be encountered
// #[derive(Clone, Debug, PartialEq)]
// pub enum Node {
//     Content(String),
//     VariableBlock(SpannedExpression),
//     Set(Set),
//     Raw(String),
//     Include(Include),
//     Block(Block),
//     Super,
//     If(If),
//     ForLoop(ForLoop),
//     // Do those need to be in the AST?
//     Continue,
//     Break,
//     FilterSection(FilterSection),
// }

#[derive(Clone, PartialEq)]
pub enum Node {
    Content(String),
    Expression(Expression),
}

impl fmt::Debug for Node {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Node::*;

        match self {
            Node::Content(s) =>  write!(f, "{:?}", s),
            Node::Expression(e) => write!(f, "{:?}", e),
        }
    }
}