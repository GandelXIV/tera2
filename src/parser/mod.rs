mod parser;

mod ast;
mod lexer;

pub use parser::Parser;

#[cfg(test)]
mod tests;
