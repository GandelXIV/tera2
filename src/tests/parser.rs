use crate::ast::Node;
use crate::parser::Parser;

#[test]
fn can_parse_ident() {
    let tests = vec![
        "{{ hello }}",
        "{{ hello_ }}",
        "{{ hello_1 }}",
        "{{ HELLO }}",
        "{{ _1 }}",
        "{{ hey.ho }}",
        "{{ h }}",
        "{{ ho }}",
        "{{ hey.ho.hu }}",
        "{{ hey.0 }}",
        "{{ h.u }}",
        "{{ hey.ho.hu }}",
        "{{ hey.0 }}",
        "{{ h.u.x.0 }}",
        "{{ hey[0] }}",
        "{{ hey[a[0]] }}",
        "{{ hey['ho'][\"hu\"] }}",
        "{{ h['u'].x[0] }}",
    ];

    for t in tests {
        println!("{:?}", t);
        let mut parser = Parser::new(t);
        parser.parse().expect("parsed failed");
        match &parser.nodes[0] {
            Node::Expression(e) => {
                assert_eq!(e.to_string(), t.replace("{{ ", "").replace(" }}", ""))
            }
            _ => unreachable!("Got something that wasn't an expression"),
        }
    }
}

#[test]
fn can_parse_expression() {
    let tests = vec![
        // literals + basic types
        ("{{ -1 }}", "-1"),
        ("{{ 1 }}", "1"),
        ("{{ 'hello' }}", "'hello'"),
        ("{{ true }}", "true"),
        ("{{ -1.2 }}", "-1.2"),
        ("{{ 1.2 }}", "1.2"),
        ("{{ a }}", "a"),
        ("{{ -a }}", "(- a)"),
        ("{{ +a }}", "(+ a)"),
        ("{{ - a * 2 }}", "(- (* a 2))"),
        ("{{ [1, 1.2, a, 'b', true] }}", "[1, 1.2, a, 'b', true]"),
        ("{{ [1, 1.2, a, 'b', true,] }}", "[1, 1.2, a, 'b', true]"), // Allows trailing `,`
        // Actual expressions
        ("{{ 1 + 2 + 3 }}", "(+ (+ 1 2) 3)"),
        ("{{ 1 + count }}", "(+ 1 count)"),
        ("{{ 1 + 2 * 3 }}", "(+ 1 (* 2 3))"),
        ("{{ a + b * c * d + e }}", "(+ (+ a (* (* b c) d)) e)"),
        // https://github.com/pallets/jinja/issues/119
        ("{{ 2 * 4 % 8 }}", "(% (* 2 4) 8)"),
        ("{{ [1 + 1, 2, 3 * 2,] }}", "[(+ 1 1), 2, (* 3 2)]"),
        // string concat
        ("{{ hey ~ ho }}", "(~ hey ho)"),
        ("{{ 1 ~ ho }}", "(~ 1 ho)"),
        ("{{ -1.2 ~ ho }}", "(~ -1.2 ho)"),
        ("{{ [] ~ ho }}", "(~ [] ho)"),
        ("{{ 'hey' ~ ho }}", "(~ 'hey' ho)"),
        ("{{ `hello` ~ ident ~ 'ho' }}", "(~ (~ 'hello' ident) 'ho')"),
        // Comparisons
        ("{{ a == b }}", "(== a b)"),
        ("{{ a != b }}", "(!= a b)"),
        ("{{ a <= b }}", "(<= a b)"),
        ("{{ a >= b }}", "(>= a b)"),
        ("{{ a < b }}", "(< a b)"),
        ("{{ a > b }}", "(> a b)"),
        ("{{ 1 + a > b }}", "(> (+ 1 a) b)"),
        ("{{ 1 + a > b * 8 }}", "(> (+ 1 a) (* b 8))"),
        // and/or
        ("{{ a and b }}", "(and a b)"),
        ("{{ a or b }}", "(or a b)"),
        (
            "{{ a + 1 == 2 or b * 3 > 10 }}",
            "(or (== (+ a 1) 2) (> (* b 3) 10))",
        ),
        // in
        ("{{ a in b }}", "(in a b)"),
        ("{{ a in b and b in c }}", "(and (in a b) (in b c))"),
        // https://github.com/mozilla/nunjucks/pull/336
        (
            "{{ msg.status in ['pending', 'confirmed'] and msg.body }}",
            "(and (in msg.status ['pending', 'confirmed']) msg.body)",
        ),
        // test
        ("{{ a is defined }}", "(is a defined)"),
        ("{{ a is not defined }}", "(not (is a defined))"),
        ("{{ a + 1 is odd }}", "(is (+ a 1) odd)"),
        ("{{ a + 1 is not odd }}", "(not (is (+ a 1) odd))"),
        ("{{ a is ending_with('s') }}", "(is a ending_with{'s'})"),
        // function calls
        (
            "{{ get_url(path=page.path, in_content=true) }}",
            "get_url{in_content=true, path=page.path}",
        ),
        ("{{ get_url() }}", "get_url{}"),
        // filters
        ("{{ a | round }}", "(| a round{})"),
        ("{{ a | round() }}", "(| a round{})"),
        ("{{ 1 + 2.1 | round }}", "(| (+ 1 2.1) round{})"),
        ("{{ [1] + [3, 2] | sort }}", "(| (+ [1] [3, 2]) sort{})"),
        ("{{ (1 + 2.1) | round }}", "(| (+ 1 2.1) round{})"),
        (
            "{{ value | json_encode | safe }}",
            "(| (| value json_encode{}) safe{})",
        ),
        (
            "{{ value | truncate(length=10) }}",
            "(| value truncate{length=10})",
        ),
        (
            "{{ get_content() | upper | safe }}",
            "(| (| get_content{} upper{}) safe{})",
        ),
        (
            "{{ admin | default or user == current_user }}",
            "(or (| admin default{}) (== user current_user))",
        ),
        (
            "{{ user == current_user or admin | default }}",
            "(or (== user current_user) (| admin default{}))",
        ),
        (
            "{{ members in interfaces | groupby(attribute='vlan') }}",
            "(in members (| interfaces groupby{attribute='vlan'}))",
        ),
        ("{{ a ~ b | upper }}", "(| (~ a b) upper{})"),
        (
            "{{ status == 'needs_restart' | ternary(truthy='restart', falsy='continue') }}",
            "(| (== status 'needs_restart') ternary{falsy='continue', truthy='restart'})",
        ),
        (
            "{{ (status == 'needs_restart') | ternary(truthy='restart', falsy='continue') }}",
            "(| (== status 'needs_restart') ternary{falsy='continue', truthy='restart'})",
        ),
        // Macro calls
        (
            "{{ macros::input(label='Name', type='text') }}",
            "macros::input{label='Name', type='text'}",
        ),
        ("{{ macros::input() | safe }}", "(| macros::input{} safe{})"),
        // Parentheses
        ("{{ ((1)) }}", "1"),
        ("{{ (2 * 3) / 10 }}", "(/ (* 2 3) 10)"),
        ("{{ (2 * 3) / 10 }}", "(/ (* 2 3) 10)"),
        // not
        ("{{ not a }}", "(not a)"),
        ("{{ not b * 1 }}", "(not (* b 1))"),
        ("{{ not a and 1 + b > 3 }}", "(and (not a) (> (+ 1 b) 3))"),
        (
            "{{ not id and not true and not 1 + c }}",
            "(and (and (not id) (not true)) (not (+ 1 c)))",
        ),
        ("{{ a not in b }}", "(not (in a b))"),
        (
            "{{ a is defined and b is not defined(1, 2) }}",
            "(and (is a defined) (not (is b defined{1, 2})))",
        ),
        (
            "{{ a is defined and not b is defined(1, 2) }}",
            "(and (is a defined) (not (is b defined{1, 2})))",
        ),
        (
            "{{ not admin | default(val=true) }}",
            "(not (| admin default{val=true}))",
        ),
    ];

    for (t, expected) in tests {
        println!("{:?}", t);
        let mut parser = Parser::new(t);
        parser.parse().expect("parsed failed");
        match &parser.nodes[0] {
            Node::Expression(e) => {
                assert_eq!(e.to_string(), expected)
            }
            _ => unreachable!("Got something that wasn't an expression"),
        }
    }
}

#[test]
fn can_parse_set() {
    let tests = vec![
        ("{% set a = 1 %}", "1"),
        ("{% set a = (b + 1) | round %}", "(| (+ b 1) round{})"),
        ("{% set a = 'hi' %}", "'hi'"),
        ("{% set a = macros::something() %}", "macros::something{}"),
        ("{% set a = utcnow() %}", "utcnow{}"),
        ("{% set a = [1, true, 'hello'] %}", "[1, true, 'hello']"),
        ("{% set_global a = 1 %}", "1"),
    ];

    for (t, expr) in tests {
        println!("{:?}", t);
        let mut parser = Parser::new(t);
        parser.parse().expect("parsed failed");
        match &parser.nodes[0] {
            Node::Set(s) => {
                assert_eq!(s.key, "a");
                assert_eq!(s.global, t.starts_with("{% set_global"));
                assert_eq!(s.value.to_string(), expr);
            }
            _ => unreachable!("Got something that wasn't a set"),
        }
    }
}

#[test]
fn can_parse_basic_include() {
    let tests = vec![
        (
            "{% include 'a.html' %}",
            Node::Include {
                files: vec!["a.html".to_string()],
                ignore_missing: false,
            },
        ),
        (
            "{% include `a.html` %}",
            Node::Include {
                files: vec!["a.html".to_string()],
                ignore_missing: false,
            },
        ),
        (
            "{% include \"a.html\" %}",
            Node::Include {
                files: vec!["a.html".to_string()],
                ignore_missing: false,
            },
        ),
        (
            "{% include \"a.html\" ignore missing %}",
            Node::Include {
                files: vec!["a.html".to_string()],
                ignore_missing: true,
            },
        ),
        (
            "{% include ['a.html', 'b.html'] %}",
            Node::Include {
                files: vec!["a.html".to_string(), "b.html".to_string()],
                ignore_missing: false,
            },
        ),
        (
            "{% include ['a.html', 'b.html'] ignore missing %}",
            Node::Include {
                files: vec!["a.html".to_string(), "b.html".to_string()],
                ignore_missing: true,
            },
        ),
    ];

    for (t, expected) in tests {
        println!("{:?}", t);
        let mut parser = Parser::new(t);
        parser.parse().expect("parsed failed");
        assert_eq!(parser.nodes[0], expected);
    }
}

#[test]
fn can_parse_extends() {
    let tests = vec![
        "{% extends 'a.html' %}",
        "{% extends `a.html` %}",
        "{% extends \"a.html\" %}",
    ];

    for t in tests {
        println!("{:?}", t);
        let mut parser = Parser::new(t);
        parser.parse().expect("parsed failed");
        match &parser.nodes[0] {
            Node::Extends(s) => {
                assert_eq!(s, "a.html");
            }
            _ => unreachable!("Got something that wasn't an include"),
        }
    }
}

#[test]
fn can_parse_comments() {
    let mut parser = Parser::new("{# a comment #}");
    parser.parse().expect("parsed failed");
    assert!(parser.nodes.is_empty());
}

#[test]
fn can_handle_whitespace_trim_left_side() {
    let tests = vec![
        (
            "\n\n{%- extends 'a.html' %}",
            Node::Extends("a.html".to_string()),
        ),
        ("Hello {{- world }}", Node::Text("Hello".to_string())),
        ("Hello {#- world #}", Node::Text("Hello".to_string())),
    ];

    for (t, expected) in tests {
        println!("{:?}", t);
        let mut parser = Parser::new(t);
        parser.parse().expect("parsed failed");
        assert_eq!(parser.nodes[0], expected);
    }
}

// #[test]
// fn can_parse_a_basic_template() {
//     let tpl = "
//     <html>
//       <head>
//         <title>{{ product.name }}</title>
//       </head>
//       <body>
//         <h1>{{ product.name }} - {{ product.manufacturer | upper }}</h1>
//         <p>{{ product.summary }}</p>
//         <p>£{{ product.price * 1.20 }} (VAT inc.)</p>
//         <p>Look at reviews from your friends {{ username }}</p>
//         <button>Buy!</button>
//       </body>
//     </html>
//     ";
//     let mut parser = Parser::new(tpl);
//     parser.parse().expect("parsed failed");
//     assert_eq!(parser.nodes[0], Node::Text("".to_string()));
// }

// TODO: Do we care about that in practice
// #[test]
// fn can_parse_expression_constant_folding() {
//     let tests = vec![
//         // https://github.com/Keats/tera/blob/master/src/parser/tests/parser.rs#L1074
//         ("{{`hello` ~ 'hey'}}", "'hellohey'"),
//         ("{{1 ~ 'ho'}}", "'1ho'"),
//         ("{{1 ~ 'ho' ~ 2}}", "'1ho2'"),
//         // comparisons
//         ("{{1 == 1}}", "true"),
//         ("{{1 == '1'}}", "false"),
//         ("{{1 == 0}}", "false"),
//     ];
//
//     for (input, expected) in tests {
//         let mut parser = Parser::new(input);
//         parser.parse().unwrap();
//         match &parser.nodes[0] {
//             Node::Expression(e) => {
//                 assert_eq!(e.to_string(), expected)
//             }
//             _ => unreachable!("Got something that wasn't an expression"),
//         }
//     }
// }
