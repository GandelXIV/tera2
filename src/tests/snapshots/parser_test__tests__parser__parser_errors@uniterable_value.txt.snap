---
source: src/tests/parser.rs
expression: res.unwrap_err()
input_file: src/tests/parser_inputs/errors/uniterable_value.txt
---
error: for loops can only iterate on strings, arrays, idents and function.
  -->   @ 1:12-1:16
